File: DL-ATM41/DL-ATM41.cs

/* https://www.decentlab.com/products/eleven-parameter-weather-station-for-lorawan */

using System;
using System.IO;
using System.Collections.Generic;

public class DecentlabDecoder
{
  private delegate double conversion(double[] x);
  public const int PROTOCOL_VERSION = 2;

  private class Sensor
  {
    internal int length { get; set; }
    internal List<SensorValue> values { get; set; }
    internal Sensor(int length, List<SensorValue> values)
    {
      this.length = length;
      this.values = values;
    }
  }

  private class SensorValue
  {
    internal string name { get; set; }
    internal string unit { get; set; }
    internal conversion convert;
    internal SensorValue(string name, string unit, conversion convert)
    {
      this.name = name;
      this.unit = unit;
      this.convert = convert;
    }
  }

  private static readonly List<Sensor> SENSORS = new List<Sensor>() {
    new Sensor(17, new List<SensorValue>() {
      new SensorValue("Solar radiation", "W⋅m⁻²", x => x[0] - 32768),
      new SensorValue("Precipitation", "mm", x => (x[1] - 32768) / 1000),
      new SensorValue("Lightning strike count", null, x => x[2] - 32768),
      new SensorValue("Lightning average distance", "km", x => x[3] - 32768),
      new SensorValue("Wind speed", "m⋅s⁻¹", x => (x[4] - 32768) / 100),
      new SensorValue("Wind direction", "°", x => (x[5] - 32768) / 10),
      new SensorValue("Maximum wind speed", "m⋅s⁻¹", x => (x[6] - 32768) / 100),
      new SensorValue("Air temperature", "°C", x => (x[7] - 32768) / 10),
      new SensorValue("Vapor pressure", "kPa", x => (x[8] - 32768) / 100),
      new SensorValue("Atmospheric pressure", "kPa", x => (x[9] - 32768) / 100),
      new SensorValue("Relative humidity", "%", x => (x[10] - 32768) / 10),
      new SensorValue("Sensor temperature (internal)", "°C", x => (x[11] - 32768) / 10),
      new SensorValue("X orientation angle", "°", x => (x[12] - 32768) / 10),
      new SensorValue("Y orientation angle", "°", x => (x[13] - 32768) / 10),
      new SensorValue("Compass heading", "°", x => x[14] - 32768),
      new SensorValue("North wind speed", "m⋅s⁻¹", x => (x[15] - 32768) / 100),
      new SensorValue("East wind speed", "m⋅s⁻¹", x => (x[16] - 32768) / 100)
    }),
    new Sensor(1, new List<SensorValue>() {
      new SensorValue("Battery voltage", "V", x => x[0] / 1000)
    })
  };

  private static int ReadInt(Stream stream)
  {
    return (stream.ReadByte() << 8) + stream.ReadByte();
  }


  public static Dictionary<string, Tuple<double, string>> Decode(byte[] msg)
  {
    return Decode(new MemoryStream(msg));
  }

  public static Dictionary<string, Tuple<double, string>> Decode(String msg)
  {
    byte[] output = new byte[msg.Length / 2];
    for (int i = 0, j = 0; i < msg.Length; i += 2, j++)
    {
      output[j] = (byte)int.Parse(msg.Substring(i, 2), System.Globalization.NumberStyles.HexNumber);
    }
    return Decode(output);
  }

  public static Dictionary<string, Tuple<double, string>> Decode(Stream msg)
  {
    var version = msg.ReadByte();
    if (version != PROTOCOL_VERSION)
    {
      throw new InvalidDataException("protocol version " + version + " doesn't match v2");
    }

    var result = new Dictionary<string, Tuple<double, string>>();
    result["Protocol version"] = new Tuple<double, string>(version, null);

    var deviceId = ReadInt(msg);
    result["Device ID"] = new Tuple<double, string>(deviceId, null);

    var flags = ReadInt(msg);
    foreach (var sensor in SENSORS)
    {
      if ((flags & 1) == 1)
      {
        double[] x = new double[sensor.length];
        for (int i = 0; i < sensor.length; i++)
        {
          x[i] = ReadInt(msg);
        }
        foreach (var val in sensor.values)
        {
          if (val.convert != null)
          {
            result[val.name] = new Tuple<double, string>(val.convert(x), val.unit);
          }
        }
      }
      flags >>= 1;
    }

    return result;
  }
}
public class Program
{
  public static void Main()
  {
    var payloads = new string[] {
      "02035a0003800a8000800080008009812b8014810880b4a57c820c810980027fe88056800880040bf5",
      "02035a00020bf5"
    };

    foreach (var pl in payloads)
    {
      var decoded = DecentlabDecoder.Decode(pl);
      foreach (var k in decoded.Keys)
      {
        Console.WriteLine(k + ": " + decoded[k]);
      }
      Console.WriteLine();
    }
  }
}

File: DL-ATM41/DL-ATM41.py
# -*- coding: utf-8 -*-

# https://www.decentlab.com/products/eleven-parameter-weather-station-for-lorawan

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import struct
from math import log
from base64 import binascii

PROTOCOL_VERSION = 2

SENSORS = [
    {'length': 17,
     'values': [{'name': 'Solar radiation',
                 'convert': lambda x: x[0] - 32768,
                 'unit': 'W⋅m⁻²'},
                {'name': 'Precipitation',
                 'convert': lambda x: (x[1] - 32768) / 1000,
                 'unit': 'mm'},
                {'name': 'Lightning strike count',
                 'convert': lambda x: x[2] - 32768},
                {'name': 'Lightning average distance',
                 'convert': lambda x: x[3] - 32768,
                 'unit': 'km'},
                {'name': 'Wind speed',
                 'convert': lambda x: (x[4] - 32768) / 100,
                 'unit': 'm⋅s⁻¹'},
                {'name': 'Wind direction',
                 'convert': lambda x: (x[5] - 32768) / 10,
                 'unit': '°'},
                {'name': 'Maximum wind speed',
                 'convert': lambda x: (x[6] - 32768) / 100,
                 'unit': 'm⋅s⁻¹'},
                {'name': 'Air temperature',
                 'convert': lambda x: (x[7] - 32768) / 10,
                 'unit': '°C'},
                {'name': 'Vapor pressure',
                 'convert': lambda x: (x[8] - 32768) / 100,
                 'unit': 'kPa'},
                {'name': 'Atmospheric pressure',
                 'convert': lambda x: (x[9] - 32768) / 100,
                 'unit': 'kPa'},
                {'name': 'Relative humidity',
                 'convert': lambda x: (x[10] - 32768) / 10,
                 'unit': '%'},
                {'name': 'Sensor temperature (internal)',
                 'convert': lambda x: (x[11] - 32768) / 10,
                 'unit': '°C'},
                {'name': 'X orientation angle',
                 'convert': lambda x: (x[12] - 32768) / 10,
                 'unit': '°'},
                {'name': 'Y orientation angle',
                 'convert': lambda x: (x[13] - 32768) / 10,
                 'unit': '°'},
                {'name': 'Compass heading',
                 'convert': lambda x: x[14] - 32768,
                 'unit': '°'},
                {'name': 'North wind speed',
                 'convert': lambda x: (x[15] - 32768) / 100,
                 'unit': 'm⋅s⁻¹'},
                {'name': 'East wind speed',
                 'convert': lambda x: (x[16] - 32768) / 100,
                 'unit': 'm⋅s⁻¹'}]},
    {'length': 1,
     'values': [{'name': 'Battery voltage',
                 'convert': lambda x: x[0] / 1000,
                 'unit': 'V'}]}
]


def decode(msg, hex=False):
    """msg: payload as one of hex string, list, or bytearray"""
    bytes_ = bytearray(binascii.a2b_hex(msg)
                       if hex
                       else msg)
    version = bytes_[0]
    if version != PROTOCOL_VERSION:
        raise ValueError("protocol version {} doesn't match v2".format(version))

    devid = struct.unpack('>H', bytes_[1:3])[0]
    bin_flags = bin(struct.unpack('>H', bytes_[3:5])[0])
    flags = bin_flags[2:].zfill(struct.calcsize('>H') * 8)[::-1]

    words = [struct.unpack('>H', bytes_[i:i + 2])[0]
             for i
             in range(5, len(bytes_), 2)]

    cur = 0
    result = {'Device ID': devid, 'Protocol version': version}
    for flag, sensor in zip(flags, SENSORS):
        if flag != '1':
            continue

        x = words[cur:cur + sensor["length"]]
        cur += sensor["length"]
        for value in sensor['values']:
            if 'convert' not in value:
                continue

            result[value['name']] = {'value': value['convert'](x),
                                     'unit': value.get('unit', None)}

    return result


if __name__ == '__main__':

    import pprint
    payloads = [
        b'02035a0003800a8000800080008009812b8014810880b4a57c820c810980027fe88056800880040bf5',
        b'02035a00020bf5',
    ]
    for pl in payloads:
        pprint.pprint(decode(pl, hex=True))
        print("")
